// openclaw/dist/plugin-sdk/subagent-registry-runtime-DjdtW15U.js
import { a as isSubagentSessionRunActive, c as resolveRequesterForChildSession, i as countPendingDescendantRunsExcludingRun, l as shouldIgnorePostCompletionAnnounceForSession, n as countActiveDescendantRuns, o as listSubagentRunsForRequester, r as countPendingDescendantRuns, s as replaceSubagentRunAfterSteer } from "./dispatch-Co2ceCdc.js";
import "./paths-Ckw4WcFk.js";
import "./model-auth-DkZtfSM9.js";
import "./logger-D7LwrCL0.js";
import "./paths-CtOdJffQ.js";
import "./proxy-env-Bkyuuf7V.js";
import "./github-copilot-token-B5aRMBY6.js";
import "./send-B5clyDuU.js";
import "./send-DYtr8_Bk.js";
import "./send-CE1lRPy-.js";
import "./deliver-BwU2fwl6.js";
import "./diagnostic-Ilr3UrDv.js";
import "./pi-model-discovery-cNATl5Eh.js";
import "./audio-transcription-runner-DWm46_W1.js";
import "./image-Bo64Bl8w.js";
import "./fetch-guard-DcmDG0QD.js";
import "./api-key-rotation-CTyrzk8e.js";
import "./local-roots-Czl62Z8b.js";
import "./proxy-fetch-BMl8EIih.js";
import "./tokens-CNK-FYqw.js";
import "./commands-registry-IGmEA0of.js";
import "./ir-RD_gYo01.js";
import "./render-Cky8fVYO.js";
import "./target-errors-DsqxSrsd.js";
import "./send-BVGdwpbt.js";
import "./outbound-attachment-DQ6JGoJC.js";
import "./fetch-B-Q2OOqY.js";
import "./send-nkF1uCCm.js";
import "./sqlite-C2qDC1oT.js";
import "./channel-activity-CLrwQV3d.js";
import "./tables-C5khgGip.js";
import "./fetch-BmOfZDIP.js";
import "./skill-commands-BpLTgyrO.js";
import "./manager-rkfLxLSi.js";
export { countActiveDescendantRuns, countPendingDescendantRuns, countPendingDescendantRunsExcludingRun, isSubagentSessionRunActive, listSubagentRunsForRequester, replaceSubagentRunAfterSteer, resolveRequesterForChildSession, shouldIgnorePostCompletionAnnounceForSession };
