// openclaw/dist/plugin-sdk/subagent-registry-runtime-CtgyXclA.js
import "./paths-aoI9lHOB.js";
import "./utils-U4csWGdS.js";
import "./paths-Dmn791zP.js";
import "./model-auth-CTzD-epg.js";
import "./target-errors-AyvbVU-y.js";
import { a as countPendingDescendantRuns, c as listSubagentRunsForRequester, d as shouldIgnorePostCompletionAnnounceForSession, i as countActiveDescendantRuns, l as replaceSubagentRunAfterSteer, o as countPendingDescendantRunsExcludingRun, s as isSubagentSessionRunActive, u as resolveRequesterForChildSession } from "./dispatch-4PYvcEOv.js";
import "./send-_-FQgtt7.js";
import "./send-WfiHbr9S.js";
import "./proxy-env-I4fYsp6I.js";
import "./github-copilot-token-B8UEBIjq.js";
import "./send-Cx455Kd4.js";
import "./fetch-guard-DUN7X0K7.js";
import "./deliver-CC4WNwqc.js";
import "./diagnostic-On3MKd62.js";
import "./pi-model-discovery-DKV4hqMA.js";
import "./audio-transcription-runner-CJ7nYvXU.js";
import "./image-2c-vkR8y.js";
import "./api-key-rotation-UhITdGL_.js";
import "./local-roots-brkwYz_C.js";
import "./proxy-fetch-CJ2KJ1Hc.js";
import "./tokens-DOkkLn5z.js";
import "./commands-registry-C9TLus2U.js";
import "./ir-C1s8-YMk.js";
import "./render-CDecAPNW.js";
import "./send-DzkJeHQb.js";
import "./outbound-attachment-fCu-0WJK.js";
import "./fetch-DdpSxwwg.js";
import "./send-DLt7YqCA.js";
import "./sqlite-DlSne7IK.js";
import "./channel-activity-Nmm3FIxp.js";
import "./tables-DjImWKeK.js";
import "./fetch-BcalEYTl.js";
import "./skill-commands-DRlejhBc.js";
import "./manager-CL4nxPdE.js";
export { countActiveDescendantRuns, countPendingDescendantRuns, countPendingDescendantRunsExcludingRun, isSubagentSessionRunActive, listSubagentRunsForRequester, replaceSubagentRunAfterSteer, resolveRequesterForChildSession, shouldIgnorePostCompletionAnnounceForSession };
