// openclaw/dist/plugin-sdk/web-Cl5v2uBr.js
import "./paths-Cxh_vBn1.js";
import { Bs as webAuthExists, Ps as pickWebChannel, js as logWebSelfId, ks as WA_WEB_AUTH_DIR } from "./sessions-C7QoPtDq.js";
import "./logger-DZpEVVp4.js";
import "./paths-D6tDENa_.js";
import "./proxy-env-BhGF_Z7k.js";
import "./send-Bi-sGJlc.js";
import "./common-Bs2Hj-Vi.js";
import "./send-BWSBxrw4.js";
import "./reply-C0BWJKME.js";
import "./github-copilot-token-CADMCozq.js";
import "./send-Dhk1bi0a.js";
import { n as monitorWebInbox, t as monitorWebChannel } from "./channel-web-BOG-ONDB.js";
import "./tokens-WCSRgKLT.js";
import "./audio-transcription-runner-C-R-bV-C.js";
import "./image-CfpEWBS6.js";
import "./fetch-guard-Y1t5rAqZ.js";
import "./api-key-rotation-BKgHzBMC.js";
import "./local-roots-CXdeCJhW.js";
import "./proxy-fetch-C-L0fJNP.js";
import "./deliver-Dr-JQ-LZ.js";
import "./commands-registry-DEF33Iqg.js";
import "./pi-model-discovery-Bv06sPuT.js";
import "./ir-CJEHVpeb.js";
import "./render-CypHTXnD.js";
import "./diagnostic-D1QFKvQB.js";
import "./send-DnEWpa_3.js";
import "./outbound-attachment-CFPN6V5o.js";
import "./fetch-DtWWTpmG.js";
import "./send-DYINXsfq2.js";
import "./sqlite-CIXWeNEz.js";
import "./channel-activity-CUM0d2E5.js";
import "./tables-VUrOoKpp.js";
import "./fetch-DmkOKs7L.js";
import "./skill-commands-Bm1cjRfc.js";
import "./manager-Ne06gS6V.js";
import { t as sendMessageWhatsApp } from "./outbound-cz8JVaMq.js";
import { i as waitForWaConnection, t as createWaSocket } from "./session-CLWv7euO.js";
import { t as loginWeb } from "./login-DDZTP059.js";
export { WA_WEB_AUTH_DIR, createWaSocket, logWebSelfId, loginWeb, monitorWebChannel, monitorWebInbox, pickWebChannel, sendMessageWhatsApp, waitForWaConnection, webAuthExists };
