// openclaw/dist/config-guard-DjfQJ2oD.js
import "./paths-BJV7vkaX.js";
import { A as theme, O as colorize, X as shouldMigrateStateFromPath, k as isRich } from "./subsystem-DBBM_0ww.js";
import { S as shortenHomePath } from "./utils-Bgkbkk_o.js";
import { Js as readConfigFileSnapshot } from "./model-selection-CGLtPxol.js";
import "./agent-scope-C2X-qS_p.js";
import "./openclaw-root-DrFjwUcG.js";
import "./logger-BefOyj1a.js";
import "./exec-Dt_VCeCB.js";
import "./github-copilot-token-D37fjdwy.js";
import { t as formatCliCommand } from "./command-format-3Z_Kl5PP.js";
import "./boolean-CJxfhBkG.js";
import "./env-BYtJCZhf.js";
import "./env-overrides-C3AZbPOQ.js";
import "./registry-BbjmejWa.js";
import "./skills-BF57gfzE.js";
import "./frontmatter-7FVJq8_7.js";
import "./plugins-DVtlo8xi.js";
import "./windows-spawn-C0ShDpd1.js";
import "./redact-DdJEHodT.js";
import "./path-alias-guards-CdGRC4Iy.js";
import "./errors-BZD05ykM.js";
import "./paths-ClV_G95V.js";
import "./chat-envelope-B16TnmTK.js";
import "./call-CehRyaj4.js";
import "./exec-approvals-allowlist-B52wcteL.js";
import "./exec-safe-bin-runtime-policy-BPVGQm2v.js";
import "./runtime-web-tools-uczAxz0g.js";
import "./plugin-auto-enable-qVhS0g_M.js";
import "./pairing-store-Bsj1iR4x.js";
import "./command-secret-targets-D2QSEv1k.js";
import "./prompt-style-_WM6kwua.js";
import "./note-50Mi-Ba7.js";
import { n as formatConfigIssueLines } from "./issue-format-Bp6X21n5.js";
import { t as loadAndMaybeMigrateDoctorConfig } from "./doctor-config-flow-CSuBEEoz.js";
//#region src/cli/program/config-guard.ts
const ALLOWED_INVALID_COMMANDS = new Set([
	"doctor",
	"logs",
	"health",
	"help",
	"status"
]);
const ALLOWED_INVALID_GATEWAY_SUBCOMMANDS = new Set([
	"status",
	"probe",
	"health",
	"discover",
	"call",
	"install",
	"uninstall",
	"start",
	"stop",
	"restart"
]);
let didRunDoctorConfigFlow = false;
let configSnapshotPromise = null;
function resetConfigGuardStateForTests() {
	didRunDoctorConfigFlow = false;
	configSnapshotPromise = null;
}
async function getConfigSnapshot() {
	if (process.env.VITEST === "true") return readConfigFileSnapshot();
	configSnapshotPromise ??= readConfigFileSnapshot();
	return configSnapshotPromise;
}
async function ensureConfigReady(params) {
	const commandPath = params.commandPath ?? [];
	if (!didRunDoctorConfigFlow && shouldMigrateStateFromPath(commandPath)) {
		didRunDoctorConfigFlow = true;
		const runDoctorConfigFlow = async () => loadAndMaybeMigrateDoctorConfig({
			options: { nonInteractive: true },
			confirm: async () => false
		});
		if (!params.suppressDoctorStdout) await runDoctorConfigFlow();
		else {
			const originalStdoutWrite = process.stdout.write.bind(process.stdout);
			const originalSuppressNotes = process.env.OPENCLAW_SUPPRESS_NOTES;
			process.stdout.write = (() => true);
			process.env.OPENCLAW_SUPPRESS_NOTES = "1";
			try {
				await runDoctorConfigFlow();
			} finally {
				process.stdout.write = originalStdoutWrite;
				if (originalSuppressNotes === void 0) delete process.env.OPENCLAW_SUPPRESS_NOTES;
				else process.env.OPENCLAW_SUPPRESS_NOTES = originalSuppressNotes;
			}
		}
	}
	const snapshot = await getConfigSnapshot();
	const commandName = commandPath[0];
	const subcommandName = commandPath[1];
	const allowInvalid = commandName ? ALLOWED_INVALID_COMMANDS.has(commandName) || commandName === "gateway" && subcommandName && ALLOWED_INVALID_GATEWAY_SUBCOMMANDS.has(subcommandName) : false;
	const issues = snapshot.exists && !snapshot.valid ? formatConfigIssueLines(snapshot.issues, "-", { normalizeRoot: true }) : [];
	const legacyIssues = snapshot.legacyIssues.length > 0 ? formatConfigIssueLines(snapshot.legacyIssues, "-") : [];
	if (!(snapshot.exists && !snapshot.valid)) return;
	const rich = isRich();
	const muted = (value) => colorize(rich, theme.muted, value);
	const error = (value) => colorize(rich, theme.error, value);
	const heading = (value) => colorize(rich, theme.heading, value);
	const commandText = (value) => colorize(rich, theme.command, value);
	params.runtime.error(heading("Config invalid"));
	params.runtime.error(`${muted("File:")} ${muted(shortenHomePath(snapshot.path))}`);
	if (issues.length > 0) {
		params.runtime.error(muted("Problem:"));
		params.runtime.error(issues.map((issue) => `  ${error(issue)}`).join("\n"));
	}
	if (legacyIssues.length > 0) {
		params.runtime.error(muted("Legacy config keys detected:"));
		params.runtime.error(legacyIssues.map((issue) => `  ${error(issue)}`).join("\n"));
	}
	params.runtime.error("");
	params.runtime.error(`${muted("Run:")} ${commandText(formatCliCommand("openclaw doctor --fix"))}`);
	if (!allowInvalid) params.runtime.exit(1);
}
const __test__ = { resetConfigGuardStateForTests };
//#endregion
export { __test__, ensureConfigReady };
