// openclaw/dist/daemon-install-DXOltKvJ.js
import "./paths-BJV7vkaX.js";
import "./subsystem-DBBM_0ww.js";
import "./utils-Bgkbkk_o.js";
import "./model-selection-CGLtPxol.js";
import "./agent-scope-C2X-qS_p.js";
import "./openclaw-root-DrFjwUcG.js";
import "./logger-BefOyj1a.js";
import "./exec-Dt_VCeCB.js";
import "./github-copilot-token-D37fjdwy.js";
import "./boolean-CJxfhBkG.js";
import "./env-BYtJCZhf.js";
import "./env-overrides-C3AZbPOQ.js";
import "./registry-BbjmejWa.js";
import "./skills-BF57gfzE.js";
import "./frontmatter-7FVJq8_7.js";
import "./plugins-DVtlo8xi.js";
import "./windows-spawn-C0ShDpd1.js";
import "./redact-DdJEHodT.js";
import "./path-alias-guards-CdGRC4Iy.js";
import "./errors-BZD05ykM.js";
import "./paths-ClV_G95V.js";
import "./chat-envelope-B16TnmTK.js";
import "./call-CehRyaj4.js";
import "./onboard-helpers-Bss2kpNZ.js";
import "./prompt-style-_WM6kwua.js";
import "./runtime-guard-B8O0vp1r.js";
import "./note-50Mi-Ba7.js";
import "./daemon-install-plan.shared-DFkAu5UR.js";
import { n as buildGatewayInstallPlan, r as gatewayInstallErrorHint, t as resolveGatewayInstallToken } from "./gateway-install-token-jnP14XtC.js";
import { r as isGatewayDaemonRuntime } from "./daemon-runtime-DM4uTR2W.js";
import { i as isSystemdUserServiceAvailable } from "./systemd-Db30ml4e.js";
import { n as resolveGatewayService } from "./service-DhSHZUzS.js";
import { n as ensureSystemdUserLingerNonInteractive } from "./systemd-linger-DohlrP1N.js";
//#region src/commands/onboard-non-interactive/local/daemon-install.ts
async function installGatewayDaemonNonInteractive(params) {
	const { opts, runtime, port } = params;
	if (!opts.installDaemon) return;
	const daemonRuntimeRaw = opts.daemonRuntime ?? "node";
	const systemdAvailable = process.platform === "linux" ? await isSystemdUserServiceAvailable() : true;
	if (process.platform === "linux" && !systemdAvailable) {
		runtime.log("Systemd user services are unavailable; skipping service install.");
		return;
	}
	if (!isGatewayDaemonRuntime(daemonRuntimeRaw)) {
		runtime.error("Invalid --daemon-runtime (use node or bun)");
		runtime.exit(1);
		return;
	}
	const service = resolveGatewayService();
	const tokenResolution = await resolveGatewayInstallToken({
		config: params.nextConfig,
		env: process.env
	});
	for (const warning of tokenResolution.warnings) runtime.log(warning);
	if (tokenResolution.unavailableReason) {
		runtime.error([
			"Gateway install blocked:",
			tokenResolution.unavailableReason,
			"Fix gateway auth config/token input and rerun onboarding."
		].join(" "));
		runtime.exit(1);
		return;
	}
	const { programArguments, workingDirectory, environment } = await buildGatewayInstallPlan({
		env: process.env,
		port,
		runtime: daemonRuntimeRaw,
		warn: (message) => runtime.log(message),
		config: params.nextConfig
	});
	try {
		await service.install({
			env: process.env,
			stdout: process.stdout,
			programArguments,
			workingDirectory,
			environment
		});
	} catch (err) {
		runtime.error(`Gateway service install failed: ${String(err)}`);
		runtime.log(gatewayInstallErrorHint());
		return;
	}
	await ensureSystemdUserLingerNonInteractive({ runtime });
}
//#endregion
export { installGatewayDaemonNonInteractive };
