// openclaw/dist/daemon-install-BtZ-wLe5.js
import "./subsystem-DVwhOlEq.js";
import "./paths-BfR2LXbA.js";
import "./boolean-BHdNsbzF.js";
import "./auth-profiles-MKCH-k1W.js";
import "./agent-scope-Cwa3GjIC.js";
import "./utils-0k7pUAq8.js";
import "./boundary-file-read-CcBHg_z-.js";
import "./logger-CDI_b4HB.js";
import "./exec-CHkvLtoM.js";
import "./github-copilot-token-CcBrBN3h.js";
import "./env-overrides-D6SjLb4J.js";
import "./version-Bxx5bg6l.js";
import "./registry-C2SN0COs.js";
import "./skills-ByMDztuV.js";
import "./frontmatter-Byi6nw9U.js";
import "./plugins-BgwnwmkA.js";
import "./windows-spawn-oQqooioe.js";
import "./redact-D9pXgQ9w.js";
import "./path-alias-guards-DdKgwxuM.js";
import "./errors-ChRw9MpT.js";
import "./paths-a7Rsgqtt.js";
import "./chat-envelope-B-k0pbMu.js";
import "./call-rN1k4f-b.js";
import "./onboard-helpers-Brn1u_2n.js";
import "./prompt-style-DoUbK5fH.js";
import "./note-D_-wdXWA.js";
import "./daemon-install-plan.shared-DdGjoRZK.js";
import "./runtime-guard-Deu-vbXy.js";
import { n as buildGatewayInstallPlan, r as gatewayInstallErrorHint, t as resolveGatewayInstallToken } from "./gateway-install-token-Bpt4OIVL.js";
import { r as isGatewayDaemonRuntime } from "./daemon-runtime-BSvRTNYU.js";
import { i as isSystemdUserServiceAvailable } from "./systemd-CcwWmJAp.js";
import { n as resolveGatewayService } from "./service-B9XdMNnP.js";
import { n as ensureSystemdUserLingerNonInteractive } from "./systemd-linger-CVhkc4zb.js";
//#region src/commands/onboard-non-interactive/local/daemon-install.ts
async function installGatewayDaemonNonInteractive(params) {
	const { opts, runtime, port } = params;
	if (!opts.installDaemon) return;
	const daemonRuntimeRaw = opts.daemonRuntime ?? "node";
	const systemdAvailable = process.platform === "linux" ? await isSystemdUserServiceAvailable() : true;
	if (process.platform === "linux" && !systemdAvailable) {
		runtime.log("Systemd user services are unavailable; skipping service install.");
		return;
	}
	if (!isGatewayDaemonRuntime(daemonRuntimeRaw)) {
		runtime.error("Invalid --daemon-runtime (use node or bun)");
		runtime.exit(1);
		return;
	}
	const service = resolveGatewayService();
	const tokenResolution = await resolveGatewayInstallToken({
		config: params.nextConfig,
		env: process.env
	});
	for (const warning of tokenResolution.warnings) runtime.log(warning);
	if (tokenResolution.unavailableReason) {
		runtime.error([
			"Gateway install blocked:",
			tokenResolution.unavailableReason,
			"Fix gateway auth config/token input and rerun onboarding."
		].join(" "));
		runtime.exit(1);
		return;
	}
	const { programArguments, workingDirectory, environment } = await buildGatewayInstallPlan({
		env: process.env,
		port,
		runtime: daemonRuntimeRaw,
		warn: (message) => runtime.log(message),
		config: params.nextConfig
	});
	try {
		await service.install({
			env: process.env,
			stdout: process.stdout,
			programArguments,
			workingDirectory,
			environment
		});
	} catch (err) {
		runtime.error(`Gateway service install failed: ${String(err)}`);
		runtime.log(gatewayInstallErrorHint());
		return;
	}
	await ensureSystemdUserLingerNonInteractive({ runtime });
}
//#endregion
export { installGatewayDaemonNonInteractive };
