// openclaw/dist/program-Bx09gyqs.js
import { A as theme, B as tryParseLogLevel, Q as hasRootVersionAlias, R as ALLOWED_LOG_LEVELS, W as getCommandPathWithRootOptions, X as hasFlag, Y as getVerboseFlag, Z as hasHelpOrVersion, k as isRich, p as defaultRuntime, w as setVerbose } from "./subsystem-DVwhOlEq.js";
import "./paths-BfR2LXbA.js";
import "./boolean-BHdNsbzF.js";
import { i as isTruthyEnvValue } from "./entry.js";
import "./auth-profiles-MKCH-k1W.js";
import { n as replaceCliName, r as resolveCliName } from "./command-format-CO3N-oGG.js";
import "./agent-scope-Cwa3GjIC.js";
import { c as escapeRegExp } from "./utils-0k7pUAq8.js";
import "./boundary-file-read-CcBHg_z-.js";
import "./logger-CDI_b4HB.js";
import "./exec-CHkvLtoM.js";
import "./github-copilot-token-CcBrBN3h.js";
import "./env-overrides-D6SjLb4J.js";
import { t as VERSION } from "./version-Bxx5bg6l.js";
import "./registry-C2SN0COs.js";
import "./skills-ByMDztuV.js";
import "./frontmatter-Byi6nw9U.js";
import "./plugins-BgwnwmkA.js";
import "./windows-spawn-oQqooioe.js";
import "./redact-D9pXgQ9w.js";
import "./path-alias-guards-DdKgwxuM.js";
import "./errors-ChRw9MpT.js";
import "./send-Do235dnE.js";
import "./send-C24Y71GV.js";
import "./compact-1mmJ_KWL.js";
import "./paths-a7Rsgqtt.js";
import "./chat-envelope-B-k0pbMu.js";
import "./models-config-B-Y0EbqA.js";
import "./tokens-T07bZqlD.js";
import "./with-timeout-BCVMjnwr.js";
import "./deliver-CVgq6pew.js";
import "./diagnostic-Dr1BHX9Y.js";
import "./send-BBhcyCkK.js";
import "./pi-model-discovery-Bf398l7Q.js";
import "./exec-approvals-allowlist-BgPONmvb.js";
import "./exec-safe-bin-runtime-policy-kbuxCUY1.js";
import "./model-catalog-DpydombN.js";
import "./fetch-C-eS-y1c.js";
import "./audio-transcription-runner-CAaIExVC.js";
import "./fetch-guard-CBQYpTN6.js";
import "./image-BYrAeHdw.js";
import "./tool-display-DKBRElit.js";
import "./api-key-rotation-Cv5-rt6-.js";
import "./proxy-fetch-mS8YHgqh.js";
import "./ir-Cc60Bzvp.js";
import "./render-Cw_ARYQd.js";
import "./target-errors-BXu-smUH.js";
import "./commands-CQ8J-rP7.js";
import "./commands-registry-CiyfJji4.js";
import "./session-cost-usage-15uZ2Vrs.js";
import "./session-utils-GZ9BN5Gx.js";
import { n as resolveCommitHash } from "./git-commit-CPPJ4l9y.js";
import "./sqlite-B831EcXc.js";
import "./call-rN1k4f-b.js";
import "./pi-tools.policy-Bq4sMN0R.js";
import "./fetch-D3omNo4J.js";
import "./pairing-store-CMMWfOlm.js";
import "./exec-approvals-Bukn_2nB.js";
import "./nodes-screen-9wihYzV3.js";
import "./restart-DqS_mqW2.js";
import "./system-run-command-CK9XrWCE.js";
import "./skill-commands-Cx5nOzJG.js";
import "./workspace-dirs-CERYhVI4.js";
import "./pairing-token-BQ0EAxXy.js";
import "./channel-activity-aUe3vPRD.js";
import "./tables-htXeeZkO.js";
import "./runtime-web-tools-_ys1IFlV.js";
import "./runtime-D26OuUH3.js";
import "./stagger-DQPqqyh6.js";
import "./channel-selection-IGFCHpWu.js";
import "./plugin-auto-enable-Bg70yf8z.js";
import "./send-AUgx_RU1.js";
import "./outbound-attachment-DP5mKTXm.js";
import "./delivery-queue-DESfaI1q.js";
import "./send-BCRbj8iN.js";
import "./fetch-BG6pJhZn.js";
import "./timeouts-Z85FZrUV.js";
import "./command-secret-targets-B4IzDSdV.js";
import "./connection-auth-Bm2DtMF6.js";
import "./onboard-helpers-Brn1u_2n.js";
import "./prompt-style-DoUbK5fH.js";
import "./pairing-labels-CadgBB0z.js";
import "./search-manager-YVuEXd-D.js";
import "./manager-mlu25MUh.js";
import "./memory-cli-0EUMX3gi.js";
import { t as formatDocsLink } from "./links-BNP-f5t1.js";
import "./cli-utils-nB6lHuak.js";
import "./help-format-BysmLRMx.js";
import "./progress-CNsW4QH6.js";
import "./plugin-registry-oCtu01sW.js";
import { n as resolveCliChannelOptions } from "./channel-options-BrcoMfN0.js";
import { t as getSubCliCommandsWithSubcommands } from "./register.subclis-Cg7SFE4d.js";
import { a as registerProgramCommands, r as getCoreCliCommandsWithSubcommands } from "./command-registry-DIN-OhIj.js";
import { r as setProgramContext } from "./program-context-BHVoTH7q.js";
import "./ports-Er5idNo8.js";
import { n as formatCliBannerLine, r as hasEmittedCliBanner, t as emitCliBanner } from "./banner-CmUCYduA.js";
import { Command, InvalidArgumentError } from "commander";
//#region src/cli/program/context.ts
function createProgramContext() {
	let cachedChannelOptions;
	const getChannelOptions = () => {
		if (cachedChannelOptions === void 0) cachedChannelOptions = resolveCliChannelOptions();
		return cachedChannelOptions;
	};
	return {
		programVersion: VERSION,
		get channelOptions() {
			return getChannelOptions();
		},
		get messageChannelOptions() {
			return getChannelOptions().join("|");
		},
		get agentChannelOptions() {
			return ["last", ...getChannelOptions()].join("|");
		}
	};
}
//#endregion
//#region src/cli/log-level-option.ts
const CLI_LOG_LEVEL_VALUES = ALLOWED_LOG_LEVELS.join("|");
function parseCliLogLevelOption(value) {
	const parsed = tryParseLogLevel(value);
	if (!parsed) throw new InvalidArgumentError(`Invalid --log-level (use ${CLI_LOG_LEVEL_VALUES})`);
	return parsed;
}
//#endregion
//#region src/cli/program/help.ts
const CLI_NAME = resolveCliName();
const CLI_NAME_PATTERN = escapeRegExp(CLI_NAME);
const ROOT_COMMANDS_WITH_SUBCOMMANDS = new Set([...getCoreCliCommandsWithSubcommands(), ...getSubCliCommandsWithSubcommands()]);
const ROOT_COMMANDS_HINT = "Hint: commands suffixed with * have subcommands. Run <command> --help for details.";
const EXAMPLES = [
	["openclaw models --help", "Show detailed help for the models command."],
	["openclaw channels login --verbose", "Link personal WhatsApp Web and show QR + connection logs."],
	["openclaw message send --target +15555550123 --message \"Hi\" --json", "Send via your web session and print JSON result."],
	["openclaw gateway --port 18789", "Run the WebSocket Gateway locally."],
	["openclaw --dev gateway", "Run a dev Gateway (isolated state/config) on ws://127.0.0.1:19001."],
	["openclaw gateway --force", "Kill anything bound to the default gateway port, then start it."],
	["openclaw gateway ...", "Gateway control via WebSocket."],
	["openclaw agent --to +15555550123 --message \"Run summary\" --deliver", "Talk directly to the agent using the Gateway; optionally send the WhatsApp reply."],
	["openclaw message send --channel telegram --target @mychat --message \"Hi\"", "Send via your Telegram bot."]
];
function configureProgramHelp(program, ctx) {
	program.name(CLI_NAME).description("").version(ctx.programVersion).option("--dev", "Dev profile: isolate state under ~/.openclaw-dev, default gateway port 19001, and shift derived ports (browser/canvas)").option("--profile <name>", "Use a named profile (isolates OPENCLAW_STATE_DIR/OPENCLAW_CONFIG_PATH under ~/.openclaw-<name>)").option("--log-level <level>", `Global log level override for file + console (${CLI_LOG_LEVEL_VALUES})`, parseCliLogLevelOption);
	program.option("--no-color", "Disable ANSI colors", false);
	program.helpOption("-h, --help", "Display help for command");
	program.helpCommand("help [command]", "Display help for command");
	program.configureHelp({
		sortSubcommands: true,
		sortOptions: true,
		optionTerm: (option) => theme.option(option.flags),
		subcommandTerm: (cmd) => {
			const hasSubcommands = cmd.parent === program && ROOT_COMMANDS_WITH_SUBCOMMANDS.has(cmd.name());
			return theme.command(hasSubcommands ? `${cmd.name()} *` : cmd.name());
		}
	});
	const formatHelpOutput = (str) => {
		let output = str;
		if (new RegExp(`^Usage:\\s+${CLI_NAME_PATTERN}\\s+\\[options\\]\\s+\\[command\\]\\s*$`, "m").test(output) && /^Commands:/m.test(output)) output = output.replace(/^Commands:/m, `Commands:\n  ${theme.muted(ROOT_COMMANDS_HINT)}`);
		return output.replace(/^Usage:/gm, theme.heading("Usage:")).replace(/^Options:/gm, theme.heading("Options:")).replace(/^Commands:/gm, theme.heading("Commands:"));
	};
	program.configureOutput({
		writeOut: (str) => {
			process.stdout.write(formatHelpOutput(str));
		},
		writeErr: (str) => {
			process.stderr.write(formatHelpOutput(str));
		},
		outputError: (str, write) => write(theme.error(str))
	});
	if (hasFlag(process.argv, "-V") || hasFlag(process.argv, "--version") || hasRootVersionAlias(process.argv)) {
		const commit = resolveCommitHash({ moduleUrl: import.meta.url });
		console.log(commit ? `OpenClaw ${ctx.programVersion} (${commit})` : `OpenClaw ${ctx.programVersion}`);
		process.exit(0);
	}
	program.addHelpText("beforeAll", () => {
		if (hasEmittedCliBanner()) return "";
		const rich = isRich();
		return `\n${formatCliBannerLine(ctx.programVersion, { richTty: rich })}\n`;
	});
	const fmtExamples = EXAMPLES.map(([cmd, desc]) => `  ${theme.command(replaceCliName(cmd, CLI_NAME))}\n    ${theme.muted(desc)}`).join("\n");
	program.addHelpText("afterAll", ({ command }) => {
		if (command !== program) return "";
		const docs = formatDocsLink("/cli", "docs.openclaw.ai/cli");
		return `\n${theme.heading("Examples:")}\n${fmtExamples}\n\n${theme.muted("Docs:")} ${docs}\n`;
	});
}
//#endregion
//#region src/cli/program/preaction.ts
function setProcessTitleForCommand(actionCommand) {
	let current = actionCommand;
	while (current.parent && current.parent.parent) current = current.parent;
	const name = current.name();
	const cliName = resolveCliName();
	if (!name || name === cliName) return;
	process.title = `${cliName}-${name}`;
}
const PLUGIN_REQUIRED_COMMANDS = new Set([
	"message",
	"channels",
	"directory",
	"agents",
	"configure",
	"onboard",
	"status",
	"health"
]);
const CONFIG_GUARD_BYPASS_COMMANDS = new Set([
	"backup",
	"doctor",
	"completion",
	"secrets"
]);
const JSON_PARSE_ONLY_COMMANDS = new Set(["config set"]);
let configGuardModulePromise;
let pluginRegistryModulePromise;
function shouldBypassConfigGuard(commandPath) {
	const [primary, secondary] = commandPath;
	if (!primary) return false;
	if (CONFIG_GUARD_BYPASS_COMMANDS.has(primary)) return true;
	if (primary === "config" && secondary === "validate") return true;
	return false;
}
function loadConfigGuardModule() {
	configGuardModulePromise ??= import("./config-guard-CkZynBXT.js").then((n) => n.t);
	return configGuardModulePromise;
}
function loadPluginRegistryModule() {
	pluginRegistryModulePromise ??= import("./plugin-registry-oCtu01sW.js").then((n) => n.n);
	return pluginRegistryModulePromise;
}
function getRootCommand(command) {
	let current = command;
	while (current.parent) current = current.parent;
	return current;
}
function getCliLogLevel(actionCommand) {
	const root = getRootCommand(actionCommand);
	if (typeof root.getOptionValueSource !== "function") return;
	if (root.getOptionValueSource("logLevel") !== "cli") return;
	const logLevel = root.opts().logLevel;
	return typeof logLevel === "string" ? logLevel : void 0;
}
function isJsonOutputMode(commandPath, argv) {
	if (!hasFlag(argv, "--json")) return false;
	const key = `${commandPath[0] ?? ""} ${commandPath[1] ?? ""}`.trim();
	if (JSON_PARSE_ONLY_COMMANDS.has(key)) return false;
	return true;
}
function registerPreActionHooks(program, programVersion) {
	program.hook("preAction", async (_thisCommand, actionCommand) => {
		setProcessTitleForCommand(actionCommand);
		const argv = process.argv;
		if (hasHelpOrVersion(argv)) return;
		const commandPath = getCommandPathWithRootOptions(argv, 2);
		if (!(isTruthyEnvValue(process.env.OPENCLAW_HIDE_BANNER) || commandPath[0] === "update" || commandPath[0] === "completion" || commandPath[0] === "plugins" && commandPath[1] === "update")) emitCliBanner(programVersion);
		const verbose = getVerboseFlag(argv, { includeDebug: true });
		setVerbose(verbose);
		const cliLogLevel = getCliLogLevel(actionCommand);
		if (cliLogLevel) process.env.OPENCLAW_LOG_LEVEL = cliLogLevel;
		if (!verbose) process.env.NODE_NO_WARNINGS ??= "1";
		if (shouldBypassConfigGuard(commandPath)) return;
		const suppressDoctorStdout = isJsonOutputMode(commandPath, argv);
		const { ensureConfigReady } = await loadConfigGuardModule();
		await ensureConfigReady({
			runtime: defaultRuntime,
			commandPath,
			...suppressDoctorStdout ? { suppressDoctorStdout: true } : {}
		});
		if (PLUGIN_REQUIRED_COMMANDS.has(commandPath[0])) {
			const { ensurePluginRegistryLoaded } = await loadPluginRegistryModule();
			ensurePluginRegistryLoaded();
		}
	});
}
//#endregion
//#region src/cli/program/build-program.ts
function buildProgram() {
	const program = new Command();
	const ctx = createProgramContext();
	const argv = process.argv;
	setProgramContext(program, ctx);
	configureProgramHelp(program, ctx);
	registerPreActionHooks(program, ctx.programVersion);
	registerProgramCommands(program, ctx, argv);
	return program;
}
//#endregion
export { buildProgram };
