// openclaw/dist/register.setup-Cv9lGj4U.js
import "./paths-BJV7vkaX.js";
import { A as theme, p as defaultRuntime } from "./subsystem-DBBM_0ww.js";
import { S as shortenHomePath } from "./utils-Bgkbkk_o.js";
import { $s as writeConfigFile, Vs as createConfigIO } from "./model-selection-CGLtPxol.js";
import { D as ensureAgentWorkspace, v as DEFAULT_AGENT_WORKSPACE_DIR } from "./agent-scope-C2X-qS_p.js";
import "./openclaw-root-DrFjwUcG.js";
import "./logger-BefOyj1a.js";
import "./exec-Dt_VCeCB.js";
import "./github-copilot-token-D37fjdwy.js";
import "./boolean-CJxfhBkG.js";
import "./env-BYtJCZhf.js";
import "./env-overrides-C3AZbPOQ.js";
import "./registry-BbjmejWa.js";
import "./skills-BF57gfzE.js";
import "./frontmatter-7FVJq8_7.js";
import "./plugins-DVtlo8xi.js";
import "./windows-spawn-C0ShDpd1.js";
import "./redact-DdJEHodT.js";
import "./path-alias-guards-CdGRC4Iy.js";
import "./errors-BZD05ykM.js";
import { o as resolveSessionTranscriptsDir } from "./paths-ClV_G95V.js";
import "./chat-envelope-B16TnmTK.js";
import "./call-CehRyaj4.js";
import "./onboard-helpers-Bss2kpNZ.js";
import "./prompt-style-_WM6kwua.js";
import { t as formatDocsLink } from "./links-O5XaN5d6.js";
import { n as runCommandWithRuntime } from "./cli-utils-CsurA6i4.js";
import "./progress-Db9dzAeG.js";
import "./runtime-guard-B8O0vp1r.js";
import { t as hasExplicitOptions } from "./command-options-DAgOYNEo.js";
import "./note-50Mi-Ba7.js";
import "./clack-prompter-D80bAXsP.js";
import "./onboarding.secret-input-KS259pS2.js";
import "./onboarding-C2TASRkC.js";
import { n as logConfigUpdated, t as formatConfigPath } from "./logging-C0o6-1t6.js";
import { t as onboardCommand } from "./onboard-DeMRz-G-.js";
import JSON5 from "json5";
import fs from "node:fs/promises";
//#region src/commands/setup.ts
async function readConfigFileRaw(configPath) {
	try {
		const raw = await fs.readFile(configPath, "utf-8");
		const parsed = JSON5.parse(raw);
		if (parsed && typeof parsed === "object") return {
			exists: true,
			parsed
		};
		return {
			exists: true,
			parsed: {}
		};
	} catch {
		return {
			exists: false,
			parsed: {}
		};
	}
}
async function setupCommand(opts, runtime = defaultRuntime) {
	const desiredWorkspace = typeof opts?.workspace === "string" && opts.workspace.trim() ? opts.workspace.trim() : void 0;
	const configPath = createConfigIO().configPath;
	const existingRaw = await readConfigFileRaw(configPath);
	const cfg = existingRaw.parsed;
	const defaults = cfg.agents?.defaults ?? {};
	const workspace = desiredWorkspace ?? defaults.workspace ?? DEFAULT_AGENT_WORKSPACE_DIR;
	const next = {
		...cfg,
		agents: {
			...cfg.agents,
			defaults: {
				...defaults,
				workspace
			}
		},
		gateway: {
			...cfg.gateway,
			mode: cfg.gateway?.mode ?? "local"
		}
	};
	if (!existingRaw.exists || defaults.workspace !== workspace || cfg.gateway?.mode !== next.gateway?.mode) {
		await writeConfigFile(next);
		if (!existingRaw.exists) runtime.log(`Wrote ${formatConfigPath(configPath)}`);
		else {
			const updates = [];
			if (defaults.workspace !== workspace) updates.push("set agents.defaults.workspace");
			if (cfg.gateway?.mode !== next.gateway?.mode) updates.push("set gateway.mode");
			logConfigUpdated(runtime, {
				path: configPath,
				suffix: updates.length > 0 ? `(${updates.join(", ")})` : void 0
			});
		}
	} else runtime.log(`Config OK: ${formatConfigPath(configPath)}`);
	const ws = await ensureAgentWorkspace({
		dir: workspace,
		ensureBootstrapFiles: !next.agents?.defaults?.skipBootstrap
	});
	runtime.log(`Workspace OK: ${shortenHomePath(ws.dir)}`);
	const sessionsDir = resolveSessionTranscriptsDir();
	await fs.mkdir(sessionsDir, { recursive: true });
	runtime.log(`Sessions OK: ${shortenHomePath(sessionsDir)}`);
}
//#endregion
//#region src/cli/program/register.setup.ts
function registerSetupCommand(program) {
	program.command("setup").description("Initialize ~/.openclaw/openclaw.json and the agent workspace").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/setup", "docs.openclaw.ai/cli/setup")}\n`).option("--workspace <dir>", "Agent workspace directory (default: ~/.openclaw/workspace; stored as agents.defaults.workspace)").option("--wizard", "Run the interactive onboarding wizard", false).option("--non-interactive", "Run the wizard without prompts", false).option("--mode <mode>", "Wizard mode: local|remote").option("--remote-url <url>", "Remote Gateway WebSocket URL").option("--remote-token <token>", "Remote Gateway token (optional)").action(async (opts, command) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			const hasWizardFlags = hasExplicitOptions(command, [
				"wizard",
				"nonInteractive",
				"mode",
				"remoteUrl",
				"remoteToken"
			]);
			if (opts.wizard || hasWizardFlags) {
				await onboardCommand({
					workspace: opts.workspace,
					nonInteractive: Boolean(opts.nonInteractive),
					mode: opts.mode,
					remoteUrl: opts.remoteUrl,
					remoteToken: opts.remoteToken
				}, defaultRuntime);
				return;
			}
			await setupCommand({ workspace: opts.workspace }, defaultRuntime);
		});
	});
}
//#endregion
export { registerSetupCommand };
