// openclaw/dist/register.agent-DHoJWl4M.js
import { A as theme, p as defaultRuntime, w as setVerbose } from "./subsystem-DVwhOlEq.js";
import "./paths-BfR2LXbA.js";
import "./boolean-BHdNsbzF.js";
import { Oo as loadConfig, or as GATEWAY_CLIENT_MODES, sr as GATEWAY_CLIENT_NAMES, tr as normalizeMessageChannel } from "./auth-profiles-MKCH-k1W.js";
import { t as formatCliCommand } from "./command-format-CO3N-oGG.js";
import { r as listAgentIds } from "./agent-scope-Cwa3GjIC.js";
import { c as normalizeAgentId } from "./session-key-GuEQvqMH.js";
import "./utils-0k7pUAq8.js";
import "./boundary-file-read-CcBHg_z-.js";
import "./logger-CDI_b4HB.js";
import "./exec-CHkvLtoM.js";
import "./github-copilot-token-CcBrBN3h.js";
import "./env-overrides-D6SjLb4J.js";
import "./version-Bxx5bg6l.js";
import "./registry-C2SN0COs.js";
import "./skills-ByMDztuV.js";
import "./frontmatter-Byi6nw9U.js";
import "./plugins-BgwnwmkA.js";
import "./windows-spawn-oQqooioe.js";
import "./redact-D9pXgQ9w.js";
import "./path-alias-guards-DdKgwxuM.js";
import "./errors-ChRw9MpT.js";
import "./send-Do235dnE.js";
import "./send-C24Y71GV.js";
import { C as agentCommand, T as resolveSessionKeyForRequest, k as createDefaultDeps } from "./compact-1mmJ_KWL.js";
import "./paths-a7Rsgqtt.js";
import "./chat-envelope-B-k0pbMu.js";
import "./models-config-B-Y0EbqA.js";
import "./tokens-T07bZqlD.js";
import "./with-timeout-BCVMjnwr.js";
import "./deliver-CVgq6pew.js";
import "./diagnostic-Dr1BHX9Y.js";
import "./send-BBhcyCkK.js";
import "./pi-model-discovery-Bf398l7Q.js";
import "./exec-approvals-allowlist-BgPONmvb.js";
import "./exec-safe-bin-runtime-policy-kbuxCUY1.js";
import "./model-catalog-DpydombN.js";
import "./fetch-C-eS-y1c.js";
import "./audio-transcription-runner-CAaIExVC.js";
import "./fetch-guard-CBQYpTN6.js";
import "./image-BYrAeHdw.js";
import "./tool-display-DKBRElit.js";
import "./api-key-rotation-Cv5-rt6-.js";
import "./proxy-fetch-mS8YHgqh.js";
import "./ir-Cc60Bzvp.js";
import "./render-Cw_ARYQd.js";
import "./target-errors-BXu-smUH.js";
import "./commands-CQ8J-rP7.js";
import "./commands-registry-CiyfJji4.js";
import "./session-cost-usage-15uZ2Vrs.js";
import "./session-utils-GZ9BN5Gx.js";
import "./sqlite-B831EcXc.js";
import { a as randomIdempotencyKey, n as callGateway } from "./call-rN1k4f-b.js";
import "./pi-tools.policy-Bq4sMN0R.js";
import "./fetch-D3omNo4J.js";
import "./pairing-store-CMMWfOlm.js";
import "./exec-approvals-Bukn_2nB.js";
import "./nodes-screen-9wihYzV3.js";
import "./restart-DqS_mqW2.js";
import "./system-run-command-CK9XrWCE.js";
import "./skill-commands-Cx5nOzJG.js";
import "./workspace-dirs-CERYhVI4.js";
import "./pairing-token-BQ0EAxXy.js";
import "./channel-activity-aUe3vPRD.js";
import "./tables-htXeeZkO.js";
import "./runtime-web-tools-_ys1IFlV.js";
import "./runtime-D26OuUH3.js";
import "./stagger-DQPqqyh6.js";
import "./channel-selection-IGFCHpWu.js";
import "./plugin-auto-enable-Bg70yf8z.js";
import "./send-AUgx_RU1.js";
import "./outbound-attachment-DP5mKTXm.js";
import "./delivery-queue-DESfaI1q.js";
import "./send-BCRbj8iN.js";
import "./fetch-BG6pJhZn.js";
import "./timeouts-Z85FZrUV.js";
import "./command-secret-targets-B4IzDSdV.js";
import "./connection-auth-Bm2DtMF6.js";
import "./onboard-helpers-Brn1u_2n.js";
import "./prompt-style-DoUbK5fH.js";
import "./pairing-labels-CadgBB0z.js";
import "./search-manager-YVuEXd-D.js";
import "./manager-mlu25MUh.js";
import "./memory-cli-0EUMX3gi.js";
import { t as formatDocsLink } from "./links-BNP-f5t1.js";
import { n as runCommandWithRuntime } from "./cli-utils-nB6lHuak.js";
import { t as formatHelpExamples } from "./help-format-BysmLRMx.js";
import { n as withProgress } from "./progress-CNsW4QH6.js";
import "./outbound-DHddprDM.js";
import "./channel-web-D-lIEz93.js";
import "./session-DNr5k_2S.js";
import "./login-DKqV1hQU.js";
import "./provider-env-vars-_PDaZmub.js";
import { t as hasExplicitOptions } from "./command-options-B6in0Ycw.js";
import "./note-D_-wdXWA.js";
import "./clack-prompter-CNzND-0H.js";
import "./config-validation-Cw-wXra8.js";
import "./agents.config-CyDFbyYG.js";
import "./enable-R4eIohap.js";
import "./install-safe-path-xC--QTS5.js";
import "./npm-pack-install-D0dAq9qx.js";
import "./skill-scanner-B8fsywKI.js";
import "./installs-B8YkUb1t.js";
import "./plugin-install-plan-CxW8TSLS.js";
import "./onboard-channels-DtgGHLbl.js";
import "./auth-choice.apply-helpers-BGNrwl8r.js";
import "./issue-format-BxaIt2aZ.js";
import { t as collectOption } from "./helpers-B9bi1Iq6.js";
import "./openai-codex-model-default-Cw_M4gpw.js";
import "./auth-token-CEK88UGz.js";
import "./logging-D9cwAGmq.js";
import "./provider-auth-helpers-Cp7doW-7.js";
import "./oauth-tls-preflight-BGauyjft.js";
import { a as agentsAddCommand, c as agentsUnbindCommand, i as agentsDeleteCommand, n as agentsListCommand, o as agentsBindCommand, r as agentsSetIdentityCommand, s as agentsBindingsCommand } from "./agents-CekOrRKq.js";
import "./auth-choice-options-DlM6VNeB.js";
import "./auth-choice-prompt-A7h5ZaYk.js";
import "./auth-choice-BvVFv1Du.js";
import "./openai-model-default-DJXmYXtw.js";
import "./model-picker-Ca2ti5TU.js";
import "./ollama-setup-BNGj9w-G.js";
//#region src/commands/agent-via-gateway.ts
const NO_GATEWAY_TIMEOUT_MS = 2147e6;
function parseTimeoutSeconds(opts) {
	const raw = opts.timeout !== void 0 ? Number.parseInt(String(opts.timeout), 10) : opts.cfg.agents?.defaults?.timeoutSeconds ?? 600;
	if (Number.isNaN(raw) || raw < 0) throw new Error("--timeout must be a non-negative integer (seconds; 0 means no timeout)");
	return raw;
}
function formatPayloadForLog(payload) {
	const lines = [];
	if (payload.text) lines.push(payload.text.trimEnd());
	const mediaUrl = typeof payload.mediaUrl === "string" && payload.mediaUrl.trim() ? payload.mediaUrl.trim() : void 0;
	const media = payload.mediaUrls ?? (mediaUrl ? [mediaUrl] : []);
	for (const url of media) lines.push(`MEDIA:${url}`);
	return lines.join("\n").trimEnd();
}
async function agentViaGatewayCommand(opts, runtime) {
	const body = (opts.message ?? "").trim();
	if (!body) throw new Error("Message (--message) is required");
	if (!opts.to && !opts.sessionId && !opts.agent) throw new Error("Pass --to <E.164>, --session-id, or --agent to choose a session");
	const cfg = loadConfig();
	const agentIdRaw = opts.agent?.trim();
	const agentId = agentIdRaw ? normalizeAgentId(agentIdRaw) : void 0;
	if (agentId) {
		if (!listAgentIds(cfg).includes(agentId)) throw new Error(`Unknown agent id "${agentIdRaw}". Use "${formatCliCommand("openclaw agents list")}" to see configured agents.`);
	}
	const timeoutSeconds = parseTimeoutSeconds({
		cfg,
		timeout: opts.timeout
	});
	const gatewayTimeoutMs = timeoutSeconds === 0 ? NO_GATEWAY_TIMEOUT_MS : Math.max(1e4, (timeoutSeconds + 30) * 1e3);
	const sessionKey = resolveSessionKeyForRequest({
		cfg,
		agentId,
		to: opts.to,
		sessionId: opts.sessionId
	}).sessionKey;
	const channel = normalizeMessageChannel(opts.channel);
	const idempotencyKey = opts.runId?.trim() || randomIdempotencyKey();
	const response = await withProgress({
		label: "Waiting for agent reply…",
		indeterminate: true,
		enabled: opts.json !== true
	}, async () => await callGateway({
		method: "agent",
		params: {
			message: body,
			agentId,
			to: opts.to,
			replyTo: opts.replyTo,
			sessionId: opts.sessionId,
			sessionKey,
			thinking: opts.thinking,
			deliver: Boolean(opts.deliver),
			channel,
			replyChannel: opts.replyChannel,
			replyAccountId: opts.replyAccount,
			bestEffortDeliver: opts.bestEffortDeliver,
			timeout: timeoutSeconds,
			lane: opts.lane,
			extraSystemPrompt: opts.extraSystemPrompt,
			idempotencyKey
		},
		expectFinal: true,
		timeoutMs: gatewayTimeoutMs,
		clientName: GATEWAY_CLIENT_NAMES.CLI,
		mode: GATEWAY_CLIENT_MODES.CLI
	}));
	if (opts.json) {
		runtime.log(JSON.stringify(response, null, 2));
		return response;
	}
	const payloads = (response?.result)?.payloads ?? [];
	if (payloads.length === 0) {
		runtime.log(response?.summary ? String(response.summary) : "No reply from agent.");
		return response;
	}
	for (const payload of payloads) {
		const out = formatPayloadForLog(payload);
		if (out) runtime.log(out);
	}
	return response;
}
async function agentCliCommand(opts, runtime, deps) {
	const localOpts = {
		...opts,
		agentId: opts.agent,
		replyAccountId: opts.replyAccount
	};
	if (opts.local === true) return await agentCommand(localOpts, runtime, deps);
	try {
		return await agentViaGatewayCommand(opts, runtime);
	} catch (err) {
		runtime.error?.(`Gateway agent failed; falling back to embedded: ${String(err)}`);
		return await agentCommand(localOpts, runtime, deps);
	}
}
//#endregion
//#region src/cli/program/register.agent.ts
function registerAgentCommands(program, args) {
	program.command("agent").description("Run an agent turn via the Gateway (use --local for embedded)").requiredOption("-m, --message <text>", "Message body for the agent").option("-t, --to <number>", "Recipient number in E.164 used to derive the session key").option("--session-id <id>", "Use an explicit session id").option("--agent <id>", "Agent id (overrides routing bindings)").option("--thinking <level>", "Thinking level: off | minimal | low | medium | high").option("--verbose <on|off>", "Persist agent verbose level for the session").option("--channel <channel>", `Delivery channel: ${args.agentChannelOptions} (omit to use the main session channel)`).option("--reply-to <target>", "Delivery target override (separate from session routing)").option("--reply-channel <channel>", "Delivery channel override (separate from routing)").option("--reply-account <id>", "Delivery account id override").option("--local", "Run the embedded agent locally (requires model provider API keys in your shell)", false).option("--deliver", "Send the agent's reply back to the selected channel", false).option("--json", "Output result as JSON", false).option("--timeout <seconds>", "Override agent command timeout (seconds, default 600 or config value)").addHelpText("after", () => `
${theme.heading("Examples:")}
${formatHelpExamples([
		["openclaw agent --to +15555550123 --message \"status update\"", "Start a new session."],
		["openclaw agent --agent ops --message \"Summarize logs\"", "Use a specific agent."],
		["openclaw agent --session-id 1234 --message \"Summarize inbox\" --thinking medium", "Target a session with explicit thinking level."],
		["openclaw agent --to +15555550123 --message \"Trace logs\" --verbose on --json", "Enable verbose logging and JSON output."],
		["openclaw agent --to +15555550123 --message \"Summon reply\" --deliver", "Deliver reply."],
		["openclaw agent --agent ops --message \"Generate report\" --deliver --reply-channel slack --reply-to \"#reports\"", "Send reply to a different channel/target."]
	])}

${theme.muted("Docs:")} ${formatDocsLink("/cli/agent", "docs.openclaw.ai/cli/agent")}`).action(async (opts) => {
		setVerbose((typeof opts.verbose === "string" ? opts.verbose.toLowerCase() : "") === "on");
		const deps = createDefaultDeps();
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentCliCommand(opts, defaultRuntime, deps);
		});
	});
	const agents = program.command("agents").description("Manage isolated agents (workspaces + auth + routing)").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/agents", "docs.openclaw.ai/cli/agents")}\n`);
	agents.command("list").description("List configured agents").option("--json", "Output JSON instead of text", false).option("--bindings", "Include routing bindings", false).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsListCommand({
				json: Boolean(opts.json),
				bindings: Boolean(opts.bindings)
			}, defaultRuntime);
		});
	});
	agents.command("bindings").description("List routing bindings").option("--agent <id>", "Filter by agent id").option("--json", "Output JSON instead of text", false).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsBindingsCommand({
				agent: opts.agent,
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.command("bind").description("Add routing bindings for an agent").option("--agent <id>", "Agent id (defaults to current default agent)").option("--bind <channel[:accountId]>", "Binding to add (repeatable). If omitted, accountId is resolved by channel defaults/hooks.", collectOption, []).option("--json", "Output JSON summary", false).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsBindCommand({
				agent: opts.agent,
				bind: Array.isArray(opts.bind) ? opts.bind : void 0,
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.command("unbind").description("Remove routing bindings for an agent").option("--agent <id>", "Agent id (defaults to current default agent)").option("--bind <channel[:accountId]>", "Binding to remove (repeatable)", collectOption, []).option("--all", "Remove all bindings for this agent", false).option("--json", "Output JSON summary", false).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsUnbindCommand({
				agent: opts.agent,
				bind: Array.isArray(opts.bind) ? opts.bind : void 0,
				all: Boolean(opts.all),
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.command("add [name]").description("Add a new isolated agent").option("--workspace <dir>", "Workspace directory for the new agent").option("--model <id>", "Model id for this agent").option("--agent-dir <dir>", "Agent state directory for this agent").option("--bind <channel[:accountId]>", "Route channel binding (repeatable)", collectOption, []).option("--non-interactive", "Disable prompts; requires --workspace", false).option("--json", "Output JSON summary", false).action(async (name, opts, command) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			const hasFlags = hasExplicitOptions(command, [
				"workspace",
				"model",
				"agentDir",
				"bind",
				"nonInteractive"
			]);
			await agentsAddCommand({
				name: typeof name === "string" ? name : void 0,
				workspace: opts.workspace,
				model: opts.model,
				agentDir: opts.agentDir,
				bind: Array.isArray(opts.bind) ? opts.bind : void 0,
				nonInteractive: Boolean(opts.nonInteractive),
				json: Boolean(opts.json)
			}, defaultRuntime, { hasFlags });
		});
	});
	agents.command("set-identity").description("Update an agent identity (name/theme/emoji/avatar)").option("--agent <id>", "Agent id to update").option("--workspace <dir>", "Workspace directory used to locate the agent + IDENTITY.md").option("--identity-file <path>", "Explicit IDENTITY.md path to read").option("--from-identity", "Read values from IDENTITY.md", false).option("--name <name>", "Identity name").option("--theme <theme>", "Identity theme").option("--emoji <emoji>", "Identity emoji").option("--avatar <value>", "Identity avatar (workspace path, http(s) URL, or data URI)").option("--json", "Output JSON summary", false).addHelpText("after", () => `
${theme.heading("Examples:")}
${formatHelpExamples([
		["openclaw agents set-identity --agent main --name \"OpenClaw\" --emoji \"🦞\"", "Set name + emoji."],
		["openclaw agents set-identity --agent main --avatar avatars/openclaw.png", "Set avatar path."],
		["openclaw agents set-identity --workspace ~/.openclaw/workspace --from-identity", "Load from IDENTITY.md."],
		["openclaw agents set-identity --identity-file ~/.openclaw/workspace/IDENTITY.md --agent main", "Use a specific IDENTITY.md."]
	])}
`).action(async (opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsSetIdentityCommand({
				agent: opts.agent,
				workspace: opts.workspace,
				identityFile: opts.identityFile,
				fromIdentity: Boolean(opts.fromIdentity),
				name: opts.name,
				theme: opts.theme,
				emoji: opts.emoji,
				avatar: opts.avatar,
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.command("delete <id>").description("Delete an agent and prune workspace/state").option("--force", "Skip confirmation", false).option("--json", "Output JSON summary", false).action(async (id, opts) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsDeleteCommand({
				id: String(id),
				force: Boolean(opts.force),
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	agents.action(async () => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			await agentsListCommand({}, defaultRuntime);
		});
	});
}
//#endregion
export { registerAgentCommands };
