// openclaw/dist/server-CjwBYRb1.js
import { t as createSubsystemLogger } from "./subsystem-DVwhOlEq.js";
import "./paths-BfR2LXbA.js";
import "./boolean-BHdNsbzF.js";
import { Lr as createBrowserRuntimeState, Oo as loadConfig, Rr as stopBrowserRuntime, Si as resolveBrowserControlAuth, Yr as resolveBrowserConfig, br as registerBrowserRoutes, vr as installBrowserAuthMiddleware, xi as ensureBrowserControlAuth, yr as installBrowserCommonMiddleware, zr as createBrowserRouteContext } from "./auth-profiles-MKCH-k1W.js";
import "./agent-scope-Cwa3GjIC.js";
import "./utils-0k7pUAq8.js";
import "./boundary-file-read-CcBHg_z-.js";
import "./logger-CDI_b4HB.js";
import "./exec-CHkvLtoM.js";
import "./github-copilot-token-CcBrBN3h.js";
import "./env-overrides-D6SjLb4J.js";
import "./version-Bxx5bg6l.js";
import "./registry-C2SN0COs.js";
import "./skills-ByMDztuV.js";
import "./frontmatter-Byi6nw9U.js";
import "./plugins-BgwnwmkA.js";
import "./windows-spawn-oQqooioe.js";
import "./redact-D9pXgQ9w.js";
import "./path-alias-guards-DdKgwxuM.js";
import "./errors-ChRw9MpT.js";
import "./paths-a7Rsgqtt.js";
import "./chat-envelope-B-k0pbMu.js";
import express from "express";
//#region src/browser/server.ts
let state = null;
const logServer = createSubsystemLogger("browser").child("server");
async function startBrowserControlServerFromConfig() {
	if (state) return state;
	const cfg = loadConfig();
	const resolved = resolveBrowserConfig(cfg.browser, cfg);
	if (!resolved.enabled) return null;
	let browserAuth = resolveBrowserControlAuth(cfg);
	let browserAuthBootstrapFailed = false;
	try {
		const ensured = await ensureBrowserControlAuth({ cfg });
		browserAuth = ensured.auth;
		if (ensured.generatedToken) logServer.info("No browser auth configured; generated gateway.auth.token automatically.");
	} catch (err) {
		logServer.warn(`failed to auto-configure browser auth: ${String(err)}`);
		browserAuthBootstrapFailed = true;
	}
	if (browserAuthBootstrapFailed && !browserAuth.token && !browserAuth.password) {
		logServer.error("browser control startup aborted: authentication bootstrap failed and no fallback auth is configured.");
		return null;
	}
	const app = express();
	installBrowserCommonMiddleware(app);
	installBrowserAuthMiddleware(app, browserAuth);
	registerBrowserRoutes(app, createBrowserRouteContext({
		getState: () => state,
		refreshConfigFromDisk: true
	}));
	const port = resolved.controlPort;
	const server = await new Promise((resolve, reject) => {
		const s = app.listen(port, "127.0.0.1", () => resolve(s));
		s.once("error", reject);
	}).catch((err) => {
		logServer.error(`openclaw browser server failed to bind 127.0.0.1:${port}: ${String(err)}`);
		return null;
	});
	if (!server) return null;
	state = await createBrowserRuntimeState({
		server,
		port,
		resolved,
		onWarn: (message) => logServer.warn(message)
	});
	const authMode = browserAuth.token ? "token" : browserAuth.password ? "password" : "off";
	logServer.info(`Browser control listening on http://127.0.0.1:${port}/ (auth=${authMode})`);
	return state;
}
async function stopBrowserControlServer() {
	await stopBrowserRuntime({
		current: state,
		getState: () => state,
		clearState: () => {
			state = null;
		},
		closeServer: true,
		onWarn: (message) => logServer.warn(message)
	});
}
//#endregion
export { startBrowserControlServerFromConfig, stopBrowserControlServer };
