// openclaw/dist/tui-cli-Cc82MqM7.js
import { A as theme, p as defaultRuntime } from "./subsystem-DVwhOlEq.js";
import "./paths-BfR2LXbA.js";
import "./boolean-BHdNsbzF.js";
import "./auth-profiles-MKCH-k1W.js";
import "./agent-scope-Cwa3GjIC.js";
import "./utils-0k7pUAq8.js";
import "./boundary-file-read-CcBHg_z-.js";
import "./logger-CDI_b4HB.js";
import "./exec-CHkvLtoM.js";
import "./github-copilot-token-CcBrBN3h.js";
import "./env-overrides-D6SjLb4J.js";
import "./version-Bxx5bg6l.js";
import "./registry-C2SN0COs.js";
import "./skills-ByMDztuV.js";
import "./frontmatter-Byi6nw9U.js";
import "./plugins-BgwnwmkA.js";
import "./windows-spawn-oQqooioe.js";
import "./redact-D9pXgQ9w.js";
import "./path-alias-guards-DdKgwxuM.js";
import "./errors-ChRw9MpT.js";
import "./paths-a7Rsgqtt.js";
import "./chat-envelope-B-k0pbMu.js";
import "./tool-display-DKBRElit.js";
import "./commands-CQ8J-rP7.js";
import "./commands-registry-CiyfJji4.js";
import "./call-rN1k4f-b.js";
import { t as parseTimeoutMs } from "./parse-timeout-BvF4wYFD.js";
import { t as formatDocsLink } from "./links-BNP-f5t1.js";
import { t as runTui } from "./tui-CiNzryOV.js";
//#region src/cli/tui-cli.ts
function registerTuiCli(program) {
	program.command("tui").description("Open a terminal UI connected to the Gateway").option("--url <url>", "Gateway WebSocket URL (defaults to gateway.remote.url when configured)").option("--token <token>", "Gateway token (if required)").option("--password <password>", "Gateway password (if required)").option("--session <key>", "Session key (default: \"main\", or \"global\" when scope is global)").option("--deliver", "Deliver assistant replies", false).option("--thinking <level>", "Thinking level override").option("--message <text>", "Send an initial message after connecting").option("--timeout-ms <ms>", "Agent timeout in ms (defaults to agents.defaults.timeoutSeconds)").option("--history-limit <n>", "History entries to load", "200").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/tui", "docs.openclaw.ai/cli/tui")}\n`).action(async (opts) => {
		try {
			const timeoutMs = parseTimeoutMs(opts.timeoutMs);
			if (opts.timeoutMs !== void 0 && timeoutMs === void 0) defaultRuntime.error(`warning: invalid --timeout-ms "${String(opts.timeoutMs)}"; ignoring`);
			const historyLimit = Number.parseInt(String(opts.historyLimit ?? "200"), 10);
			await runTui({
				url: opts.url,
				token: opts.token,
				password: opts.password,
				session: opts.session,
				deliver: Boolean(opts.deliver),
				thinking: opts.thinking,
				message: opts.message,
				timeoutMs,
				historyLimit: Number.isNaN(historyLimit) ? void 0 : historyLimit
			});
		} catch (err) {
			defaultRuntime.error(String(err));
			defaultRuntime.exit(1);
		}
	});
}
//#endregion
export { registerTuiCli };
